src/hotspot/share/utilities/nonblockingQueue.inline.hpp

```diff
@@ -181,17 +190,21 @@ bool NonblockingQueue<T, next_ptr>::try_pop(T** node_ptr) {
     // Any further try_pops will consider the queue empty until a
     // push/append completes by installing a new head.
 
-    // Attempt to change the queue tail from result to NULL.  Failure of the
-    // cmpxchg indicates that a concurrent push/append updated the tail first.
-    // That operation will eventually recognize the old tail (our result) is
-    // no longer in the list and update head from the list being appended.
-    Atomic::cmpxchg(&_tail, result, (T*)NULL);
+    // The order of the two cmpxchgs doesn't matter algorithmically, but
+    // dealing with _head first gives a stronger invariant in append, and is
+    // also consistent with [Clause 1b].
 
     // Attempt to change the queue head from result to NULL.  Failure of the
     // cmpxchg indicates a concurrent operation updated _head first.  That
-    // could be either a push/extend or a try_pop in [Clause 1b].
+    // could be either a push/append or a try_pop in [Clause 1b].
     Atomic::cmpxchg(&_head, result, (T*)NULL);
```

Comment:
These `NULL`s could be replaced by `nullptr`.

Reply:
I'm planning to do that, since this code has been going through a lot of recent churn anyway.  But I didn't want to mix that cleanup with functional changes.